src/context.rs

```diff
@@ -1,88 +1,360 @@
-    pub x_span_id: Option<String>,
-
-    /// Authorization data, filled in from middlewares.
-    pub authorization: Option<Authorization>,
-    /// Raw authentication data, for use in making HTTP requests as a client.
-    pub auth_data: Option<AuthData>,
-    logger: Option<slog::Logger>,
+use std::marker::Sized;
+use super::XSpanIdString;
+
+/// Defines methods for accessing, modifying, adding and removing the data stored
+/// in a context. Used to specify the requirements that a hyper service makes on
+/// a generic context type that it receives with a request, e.g.
+///
+/// ```rust
+/// # extern crate hyper;
+/// # extern crate swagger;
+/// # extern crate futures;
+/// #
+/// # use swagger::context::*;
+/// # use futures::future::{Future, ok};
+/// # use std::marker::PhantomData;
+/// #
+/// # struct MyItem;
+/// # fn do_something_with_my_item(item: &MyItem) {}
+/// #
+/// struct MyService<C> {
+///     marker: PhantomData<C>,
+/// }
+///
+/// impl<C> hyper::server::Service for MyService<C>
+///     where C: Has<MyItem>,
+/// {
+///     type Request = (hyper::Request, C);
+///     type Response = hyper::Response;
+///     type Error = hyper::Error;
+///     type Future = Box<Future<Item=Self::Response, Error=Self::Error>>;
+///     fn call(&self, (req, context) : Self::Request) -> Self::Future {
+///         do_something_with_my_item(Has::<MyItem>::get(&context));
+///         Box::new(ok(hyper::Response::new()))
+///     }
+/// }
+///
+/// # fn main() {}
+/// ```
+pub trait Has<T> {
+    /// Get an immutable reference to the value.
+    fn get(&self) -> &T;
+    /// Split into a the value and the remainder.
+    fn get_mut(&mut self) -> &mut T;
+    /// Set the value.
+    fn set(&mut self, value: T);
 }
 
-/// Trait for retrieving a logger from a struct.
-pub trait HasLogger {
-    /// Retrieve the context logger
-    fn logger(&self) -> &Option<slog::Logger>;
+/// Defines a method for permanently extracting a value, changing the resulting
+/// type. Used to specify that a hyper service consumes some data from the context,
+/// making it unavailable to later layers, e.g.
+///
+/// ```rust
+/// # extern crate hyper;
+/// # extern crate swagger;
+/// # extern crate futures;
+/// #
+/// # use swagger::context::*;
+/// # use futures::future::{Future, ok};
+/// # use std::marker::PhantomData;
+/// #
+/// # struct MyItem;
+/// # fn do_something_with_my_item(item: &MyItem) {}
+/// #
+/// struct MiddlewareService<T, C> {
+///     inner: T,
+///     marker: PhantomData<C>,
+/// }
+///
+/// impl<T, C> hyper::server::Service for MiddlewareService<T, C>
+///     where
+///         C: Pop<MyItem>,
+///         T: hyper::server::Service<Request = (hyper::Request, C::Result)>
+/// {
+///     type Request = (hyper::Request, C);
+///     type Response = T::Response;
+///     type Error = T::Error;
+///     type Future = T::Future;
+///     fn call(&self, (req, context) : Self::Request) -> Self::Future {
+///         let (item, context) = context.pop();
```

Comment:
Is it possible to pop multiple items from a context? How do you specify that?

Reply:
Yes, you can have arbitrary chains of pops and pushes. You add a constraint that the `Result` type of each trait implements the next trait in the chain. E.g.
```
impl<T, C, D, E> Service for MiddlewareService<T, C>
where
    C: Pop<Type1, Result=D>,
    D: Pop<Type2, Result=E>,
    E : Pop<Type3>,
    T: Service<Request=(Request, E::Result)>
{
    // other associated types here
    type Request = (Request, C);
    fn call(&self, (req, context): Self::Request) -> Self::Future {
        let (_, context) : (_, D)= context.pop();
        let (_, context) : (_, E)= context.pop();
        let (_, context) : (_, E::Result)= context.pop();
        self.inner.call((req, context))
    }
}
```

Reply:
Cool, I think that might be worth an example somewhere